### PR TITLE
Rust YJIT

### PR DESCRIPTION
**Pull request by XrXr**
In December 2021, we opened an [issue] to solicit feedback regarding the porting of the YJIT codebase from C99 to Rust. There were some reservations, but this project was given the go ahead by Ruby core developers and Matz. Since then, we have successfully completed the port of YJIT to Rust. We are opening this pull request to upstream Rust YJIT, effectively replacing the C version of YJIT.

The new Rust version of YJIT has reached parity with the C version, in that it passes all the CRuby tests, is able to run all of the YJIT benchmarks, and performs similarly to the C version (because it works the same way and largely generates the same machine code). We've even incorporated some design improvements, such as a more fine-grained constant invalidation mechanism which we expect will make a big difference in Ruby on Rails applications.

Because we want to be careful, YJIT is guarded behind a configure option:

```shell
./configure --enable-yjit # Build YJIT in release mode
./configure --enable-yjit=dev # Build YJIT in dev/debug mode
```

By default, YJIT does not get compiled and cargo/rustc is not required. If YJIT is built in dev mode, then `cargo` is used to fetch development dependencies, but when building in release, `cargo` is not required, only `rustc`. At the moment YJIT requires Rust 1.60.0 or newer.

The YJIT command-line options remain mostly unchanged, and more details about the build process are documented in `doc/yjit/yjit.md`.

The CI tests have been updated and do not take any more resources than before.

The development history of the Rust port is available at the following commit for interested parties: https://github.com/Shopify/ruby/commit/958c89d5ddb5d98cabcd2a9962ea42e4c9194728

Our hope is that Rust YJIT will be compiled and included as a part of system packages and compiled binaries of the Ruby 3.2 release. We do not anticipate any major problems as Rust is well supported on every platform which YJIT supports, but to make sure that this process works smoothly, we would like to reach out to those who take care of building systems packages before the 3.2 release is shipped and resolve any issues that may come up.

---

Please feel free to comment or ask questions on this pull request if you have any.

_AppVeyor_ CI checks are currently red on the master branch, so they will be red on this PR as well.

[issue]: https://bugs.ruby-lang.org/issues/18481

### CODE REVIEW

File: yjit/src/asm/mod.rs
Comment (Stranger6667) [inline yjit/src/asm/mod.rs:175]: The unconditional allocation could be avoided with `String::as_str`, so the string is only allocated when the comment is not the same:

```suggestion
            if this_line_comments.last().map(String::as_str) != Some(comment) {
                this_line_comments.push(String::from(comment));
            }
```

File: yjit/src/asm/mod.rs
Comment (Stranger6667) [inline yjit/src/asm/mod.rs:169]: It could be done with with a single `BTreeMap` lookup:

```suggestion
            let this_line_comments = self.asm_comments.entry(cur_ptr).or_insert_with(Vec::new);
```

File: defs/gmake.mk
Comment (nobu) [inline defs/gmake.mk:366]: You can split `if` and `else` targets below by

```
yjit-static-lib: yjit-static-lib-$(YJIT_SUPPORT)
yjit-static-lib-release:
	echo 'building Rust YJIT (release mode)'
	$(RUSTC)...
yjit-static-lib-dev: 
	echo 'building Rust YJIT (dev mode)'
	$(RUSTC)...
```

File: template/Makefile.in
Comment (nobu) [inline template/Makefile.in:306]: I'm uncertain if all `make` accept a comment in middle of a recipe.

File: template/Makefile.in
Comment (nobu) [inline template/Makefile.in:315]: Why not simply include `$(YJIT_LIBS)` in `$(LIBRUBY_A_OBJS)` or `$(INITOBJS)`?

File: yjit/src/cruby.rs
Comment (ianks) [inline yjit/src/cruby.rs:337]: :+1:

For those interested, [the docs say this](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#isize-and-usize):


> The isize and usize types are pointer-sized signed and unsigned integers. They have the same layout as the [pointer types](https://rust-lang.github.io/unsafe-code-guidelines/layout/pointers.html) for which the pointee is Sized, and are layout compatible with C's uintptr_t and intptr_t types

File: defs/gmake.mk
Comment (ianks) [inline defs/gmake.mk:373]: Might not be in scope for this PR, but would we need to account for the cargo build target here when cross compiling ruby?

Since yjit is currently x86_64, I think these mappings should work:

| ruby | rust |
|------|------|
| x86_64-linux | x86_64-unknown-linux-gnu |
| x86_64-darwin | x86_64-apple-darwin |
| x64-mingw32 | x86_64-pc-windows-gnu |
| x64-mingw-ucrt | x86_64-pc-windows-gnu |

File: doc/yjit/yjit.md
Comment (ianks) [inline doc/yjit/yjit.md:59]: Since 2021 edition is required, we should specify that. Not sure how up to date rustc is for the various package managers

```suggestion
- The Rust compiler `rustc` (>= 1.56.0
) and Cargo (if you want to build in dev/debug mode)
```

File: yjit.c
Comment (bjorn3) [inline yjit.c:56]: Rustc doesn't do `-fvisibility=hidden`. Instead when linking a cdylib it uses a version script to hide exported symbols. This is because any library may end up being linked into a rust dylib in which case all symbols need to be exported for other crates to depend on. In any case you can only get symbol conflicts for symbols not marked as `#[no_mangle]` when using the exact same rustc version as the rustc version is hashed into mangled symbols. To further reduce the chance of conflicts you can pass a unique string to `-Cmetadata` which is hashed into mangled symbols too. That leaves the standard library of which most symbols are mangled and thus would only conflict when the exact same version is used which is probably fine. There are a couple of symbols in the standard library that aren't mangled though because of cyclic dependencies between libcore, liballoc and libstd. For them changing rustc to mark them as protected would be fine I think. In addition you could use a version script to prevent exporting any symbols starting with `__rust_`. This is true for the majority of the unmangled symbols. Only a couple use a different prefix (`rust_eh_personality`, `__rg_alloc`, `__rdl_alloc`, ...). Arguably they should be changed to use the `__rust_` prefix too. In addition maybe rustc could start including the rustc version or a hash in them? In any case most of them haven't been changing a lot between versions and may thus be safe to export until rustc gets changed for better hygiene.

File: yjit/bindgen/src/main.rs
Comment (bjorn3) [inline yjit/bindgen/src/main.rs:6]: nit: Using `extern crate` is no longer necessary since the 2018 edition in the majority of the cases, including this case.

File: yjit/src/lib.rs
Comment (bjorn3) [inline yjit/src/lib.rs:1]: Is this still necessary?

File: yjit/Cargo.toml
Comment (Stranger6667) [inline yjit/Cargo.toml:9]: The OP comment mentions `1.60.0`

> At the moment YJIT requires Rust 1.60.0 or newer.

File: yjit/Cargo.toml
Comment (Stranger6667) [inline yjit/Cargo.toml:26]: nit: Probably a typo, as the feature below is called `disasm`, not `disassembly`

File: yjit/src/disasm.rs
Comment (Stranger6667) [inline yjit/src/disasm.rs:84]: As `format!` allocates, it might be better to use [write!](https://doc.rust-lang.org/std/macro.write.html) instead as `String` implements `std::fmt::Write` (`write!` [can't fail](https://github.com/rust-lang/rust/blob/1.47.0/library/alloc/src/string.rs#L2414-L2427), so `unwrap` is safe):

```suggestion
    write!(&mut out, "NUM BLOCK VERSIONS: {}\n", block_list.len()).unwrap();
```

File: yjit.c
Comment (bjorn3) [reply yjit.c:56]: I have added the rustc changes to my personal TODO list. Unless someone else makes these changes first it will probably take a while for me to get to it though.

File: yjit/src/options.rs
Comment (RossSmyth) [inline yjit/src/options.rs:41]: I'd consider using something else other than a static mut for holding on to options. Especially a pub one. Is there any reason for this to be public? 
rust-lang/rust#53639

**Comment by sbeckeriv** [comment]
Dearest Maintainer, 

I am very excited about this change! I checked out the branch and ran clippy (https://github.com/rust-lang/rust-clippy) `cargo +nightly clippy --fix -Z unstable-options --allow-dirty`  the auto fix removed un-needed returns, clones, and allocations. There are 54 other warnings some of which are helpful for performance. If a clean clippy run could be had it would be nice to enforce in build process. 

Thank you for you work on this! I am happy to leave each clippy suggestion as a github suggestion if that helps.

Becker

File: yjit/src/options.rs
Comment (sifton) [reply yjit/src/options.rs:70]: i suppose to preclude accidental reuse of `c_str` without null checking done by `opt_str` you could simplify the first two lines as follows:

`let opt_str: &str = unsafe { CStr::from_ptr(str_ptr) }.to_str().ok()?;`

**Comment by ianks** [comment]
> Dearest Maintainer,
> 
> I am very excited about this change! I checked out the branch and ran clippy (https://github.com/rust-lang/rust-clippy) `cargo +nightly clippy --fix -Z unstable-options --allow-dirty` the auto fix removed un-needed returns, clones, and allocations. There are 54 other warnings some of which are helpful for performance. If a clean clippy run could be had it would be nice to enforce in build process.
> 
> Thank you for you work on this! I am happy to leave each clippy suggestion as a github suggestion if that helps.
> 
> Becker

Good call! Might be worthwhile to add add [a clippy check](https://github.com/actions-rs/clippy-check) to CI soon so the lints don’t pile up. Could help with adoption and keeping up with best practices

File: yjit/src/asm/mod.rs
Comment (nbdd0121) [reply yjit/src/asm/mod.rs:175]: `.map(String::as_str)` can also be written as `.as_deref()`

File: yjit/bindgen/src/main.rs
Comment (coolreader18) [reply yjit/bindgen/src/main.rs:279]: That's not true, extend() for PathBuf is equivalent to `for x in it { p.push(x) }`, but you *can* call push() with a path that has separators in it. `out_path.push("yjit/src/cruby_bindings.inc.rs")` is fine and cross-platform, and it can actually just be `let out_path = src_root.join("yjit/src/cruby_bindings.inc.rs");` at that point.

File: yjit/src/asm/mod.rs
Comment (Lucretiel) [inline yjit/src/asm/mod.rs:30]: Might it be possible to do `From<NonNull<u8>>` instead? `NonNull<T>` is a `*mut T` that's guaranteed to not be null, and `Option<NonNull<T>>` is guaranteed to be FFI compatible with `*mut T`

File: yjit/src/options.rs
Comment (Lucretiel) [reply yjit/src/options.rs:70]: Sorry, to be clear, I think they meant we need to check for null pointers here, not null termination. `CStr::from_ptr` must be called with a non null pointer (perhaps `parse_option` could accept a `NonNull<c_char>`?)

Comment (sifton) [reply yjit/src/options.rs:70]: ahh, good catch! yeah, that makes sense.

File: yjit/src/disasm.rs
Comment (Yam76) [inline yjit/src/disasm.rs:122]: Can this check every block be avoided? Maybe by popping off the end of the `block_list` iterator and handling it separately?

File: yjit/src/options.rs
Comment (RossSmyth) [reply yjit/src/options.rs:70]: Oh yeah sorry I wasn't clear, Lucretiel is correct in that I meant a null pointer being input rather than a null sentinel. I would also be careful with just raw Nonnull since this is a pub function that is called directly from C FFI. 

What I would consider "idiomatic" would be having the C FFI code do the conversions to Rust happy types. This is just a quick example but this is approximately how I would do it. But obviously the choice is yours.
```rust
use std::os::raw::c_char;
use std::ptr::NonNull;
use std::ffi::CStr;

// I would have it as an optional non-null since this is pub so this is no guarantee that someone will provide a null.
// I wouldn't change unless this conversion turns out to be a performance bottleneck. If it does, then putting an 
// `.unwrap_unchecked()` would be the move cause it still expresses intent.
pub extern "C" fn rb_yjit_parse_option(str_ptr: Option<NonNull<raw::c_char>>) -> bool
{
    match str_ptr {
        Some(str_ptr) => {
            let new_str = unsafe { 
                CStr::from_ptr(str_ptr.as_ptr()).to_str()
            };
            match new_str {
                Ok(rust_str) => parse_option(rust_str).is_some(),
                Err(_) => false
            }
        },
        None => false
    }
}
```

**Comment by RossSmyth** [comment]
I would consider setting up a rustfmt configuration as the formatting is not very consistent.

File: yjit/src/disasm.rs
Comment (Yam76) [inline yjit/src/disasm.rs:133]: Minor thing, can this check every iteration be avoided, maybe by popping off the end of the `block_list` iterator and handling it separately?

**Comment by RossSmyth** [comment]
I do not know much about JITs and was just linked here by a friend. This is very impressive. There a few points that I just want to make sure are known from the view of a Rust dev.:

- `static mut` is a huge footgun waiting. I mentioned it above, but if this crate ever plans to use mutithreading a large part of this crate will be trivially UB. Even without multithreading UB is quite easy because the guarantees that references assert are extremely strong, and `static mut` makes it very easy to accidentally mess this up. This is done by modifying the `static mut T` while a `&mut T` to it is still live, leading to the program being UB. This thread is a good read on the reasons: rust-lang/rust#53639. The TL;DR is to use a `static UnsafeCell<T>` and raw pointers for the most part. 
- C FFI vs. Rust 
The general recommended way to go about C FFI is to write the Rust functions without considering the C interface, and then make a separate crate that does the logic for conversions/validating things between C and Rust, while this code has a lot of things just in extern C functions. Though I understand why this isn't done for the most part, but since some do use a pattern similar, separating the FFI interface from the main logic, I would consider it.
- Unsafe macros
This is personal preference, but I prefer to leave unsafe explicit rather than wrapping in macros so it's more easily auditable later.

I may try running this with Miri for laughs later. Anyways thanks for the cool project! I'll be keeping an eye on it.

**Comment by XrXr** [comment]
Thanks for all the suggestions! I pushed some commits to adopt some small changes, but please understand that I simply can’t get to all of them in this PR, especially the suggestions that generate big diffs. I’ll try to respond to your comments, but sorry in advance if I don’t get to yours. I’m outnumbered and I’m not very good or quick with words. :pray:

Since we use linear history in this repo I would like to keep the commit set in this PR small to make the merge manageable. After the initial merge of course we’re open to adopting improvements! Regarding clippy/rustfmt specifically, we’ll think about setting something up because it looks like a lot of folks are interested.

File: defs/gmake.mk
Comment (XrXr) [reply defs/gmake.mk]: Thanks for this nice technique; it makes `make V=1` nicer! I added f04da3cf8798e25885a2383737b1e40a6d0cbeb3

File: yjit.c
Comment (XrXr) [reply yjit.c:56]: Thanks for the context around `-fvisibility=hidden`, and for working on rustc! I _think_ risk of collision is acceptably low for now since the Rust code is built as a static lib and linked into a DSO for that configuration, and the linker takes care of only including parts of the Rust stdlib that we use? I haven't verified that, though, and I added that to my TODO list. Thanks!

Also I think I understand the rationale mentioned in the comment a bit better now. It looks like `-fvisibility=hidden` is only relevant for dynamic libs so the prefixes are for avoiding collisions in the static lib case. (The comment was about practices in the C parts of Ruby)

File: template/Makefile.in
Comment (XrXr) [reply template/Makefile.in:316]: I did try that, but that creates a nested archive. I got some build errors since `ld` has trouble finding YJIT symbols when dealing with nested archives.

File: template/Makefile.in
Comment (XrXr) [reply template/Makefile.in]: Yeah, that was a bit risky. Addressed in f04da3cf8798e25885a2383737b1e40a6d0cbeb3

**Comment by XrXr** [comment]
> Should I understand the Rust YJIT code to be under the same license as the rest of the Ruby codebase?

Right, we use the same license, no change from when YJIT is in C. I'm not a lawyer though!

> I'm interested in potentially using the dynamic assembler (and possibly more) in my own project.

You are welcome to do that, but I feel there has got to be better alternatives on crates.io that's more convenient to use and more Rusty. Good library design is hard work, and Rust YJIT doesn't do that work because, well, we don't intend to be a Rust library on crates.io! Just be aware.

File: yjit/src/disasm.rs
Comment (XrXr) [reply yjit/src/disasm.rs:84]: Today I learned! I'm going to keep the current code though since we don't need disassembly to be super fast and `format!` shows up in intro books so it's more beginner friendly.

File: yjit/src/asm/mod.rs
Comment (XrXr) [reply yjit/src/asm/mod.rs]: Thanks! It makes this function much more concise. I ended up going with `or_default()` in de6a18a00694548184bb4b04810fbf2e4ff813d1

File: yjit/src/asm/mod.rs
Comment (XrXr) [reply yjit/src/asm/mod.rs]: I went with `String::as_str` in de6a18a00694548184bb4b04810fbf2e4ff813d1 since the `.as_deref()` feels harder to trace through in the docs because of the extra indirection and needing to understand the `Deref` trait. Thank you both!

File: yjit/src/lib.rs
Comment (XrXr) [reply yjit/src/lib.rs:1]: Yeah, still needed for now. We've accumulated a bunch of these and will need to do a pass to address them.

File: yjit/bindgen/src/main.rs
Comment (XrXr) [reply yjit/bindgen/src/main.rs:278]: Ayy the current code looks fine to me. We don't need this to be fast and it looks similar to what I find in the docs.

**Comment by ianks** [comment]
> I would consider setting up a rustfmt configuration as the formatting is not very consistent.

As annoying as the initial diff might be, I’m very +1 on doing this as early as possible. I suspect a large majority of rust devs have this configured as part of their workflow / editor configuration, so enabling it will reduce the burden of contributing.

If it’s not in this PR, a follow up formatting PR + a quick-merge would work as well!

File: template/Makefile.in
Comment (ianks) [reply template/Makefile.in:316]: Was messing around on x86_64-darwin, and this patch seems to work. Does this seem reasonable? If so, I will throw up a PR to see if we can make it pass on all platforms

```diff
diff --git a/template/Makefile.in b/template/Makefile.in
index fee62c8d30..de339847a0 100644
--- a/template/Makefile.in
+++ b/template/Makefile.in
@@ -304,16 +304,7 @@ $(LIBRUBY_A):
 		@$(RM) $@
 		@-[ -z "$(EXTSTATIC)" ] || $(PRE_LIBRUBY_UPDATE)
 		$(ECHO) linking static-library $@
-		$(Q) $(AR) $(ARFLAGS) $@ $(LIBRUBY_A_OBJS) $(INITOBJS)
-		$(Q) if [ -f '$(YJIT_LIBS)' ]; then \
-		  set -eu && \
-		  echo 'merging $(YJIT_LIBS) into $@' && \
-		  $(RMALL)    '$(CARGO_TARGET_DIR)/libyjit/' && \
-		  $(MAKEDIRS) '$(CARGO_TARGET_DIR)/libyjit/' && \
-		  $(CP) '$(YJIT_LIBS)' '$(CARGO_TARGET_DIR)/libyjit/' && \
-		  (cd '$(CARGO_TARGET_DIR)/libyjit/' && $(AR) -x libyjit.a) && \
-		  $(AR) $(ARFLAGS) $@ $$(find '$(CARGO_TARGET_DIR)/libyjit/' -name '*.o') ; \
-		fi
+		$(Q) $(AR) $(ARFLAGS) $@ $(LIBRUBY_A_OBJS) $(INITOBJS) $(YJIT_LIBS)
 		@-$(RANLIB) $@ 2> /dev/null || true
 
 verify-static-library: $(LIBRUBY_A)
```

**Review by maximecb — APPROVED** [review]

File: yjit/src/disasm.rs
Comment (nbdd0121) [reply yjit/src/disasm.rs:84]: I don't think it's good to keep current code -- Clippy will start to warn about this (https://rust-lang.github.io/rust-clippy/master/#format_push_string) since 1.61.0.

File: yjit/src/cruby.rs
Comment (XrXr) [reply yjit/src/cruby.rs:609]: Noted! Clearly this won't build on Windows as is. Might want to do `.try_into()?` or something when we tackle Windows support.

Comment (clemenswasser) [reply yjit/src/cruby.rs:609]: Sure, sounds even better 👍

File: yjit/src/asm/mod.rs
Comment (saethlin) [inline yjit/src/asm/mod.rs:46]: FYI this is a dereference of a null pointer, which is UB. It probably doesn't codegen to an actual read of address 0, so you don't get a fault, but it's UB nonetheless. If possible, it would be good to use the `memoffset` crate instead.

I don't know if this was copy-pasted from elsewhere, but `bindgen` has the same UB code pattern in it. I wouldn't be surprised if this code has spread all over.

Comment (maximecb) [reply yjit/src/asm/mod.rs:46]: LLVM definitely doesn't read an address of 0 because this is an address computation, no memory is actually being read.

We've been trying to avoid using external crates which is the problem. I would like to use some sort of built in `offsetof` operator instead, it would look cleaner, but unfortunately it's only accessible through an external crate.

Comment (saethlin) [reply yjit/src/asm/mod.rs:46]: While I greatly worry about pasting `unsafe` code from external crates (multiple times, I've fixed the same piece of UB in 2 or 3 crates because they all copy+pasted instead of having a dependency), `memoffset` is MIT and the source for `offset_of!` is right here: https://github.com/Gilnaa/memoffset/blob/66f4c956a72daabeafa87ba9aef339d9035d78b4/src/offset_of.rs#L90 and I think it would be better to copy+paste than definitely have UB.

Comment (XrXr) [reply yjit/src/asm/mod.rs:46]: Thanks for letting us know! This macro is actually unused and I'm in the process of deleting it while addressing other dead code lints. We have `#[allow(unused_macros)]` at the moment which hides the lint. We'll be sure to try for a UB free implementation when we need one. :)

**Pull request merged**
